tauri-updater/src/lib.rs

```diff
@@ -1,19 +1,1034 @@
+  pub async fn download_and_install(&self, pub_key: Option<String>) -> Result {
+    // get OS
+    let target = self.target.clone();
+    // download url for selected release
+    let url = self.download_url.clone();
+    // extract path
+    let extract_path = self.extract_path.clone();
+
+    // make sure we NEED to install it ...
+
+    // make sure we can install the update on linux
+    // We fail here because later we can add more linux support
+    // actually if we use APPIMAGE, our extract path should already
+    // be set with our APPIMAGE env variable, we don't need to do
+    // anythin with it yet
+    if target == "linux" && env::var_os("APPIMAGE").is_none() {
+      bail!(
+        Error::Config,
+        "APPIMAGE env is not defined -- updates are not supported."
+      )
+    }
+
+    // used  for temp file name
+    // if we cant extract app name, we use unix epoch duration
+    let current_time = SystemTime::now()
+      .duration_since(UNIX_EPOCH)
+      .expect("Unable to get Unix Epoch")
+      .subsec_nanos()
+      .to_string();
+
+    // get the current app name
+    let bin_name = std::env::current_exe()
+      .ok()
+      .and_then(|pb| pb.file_name().map(|s| s.to_os_string()))
+      .and_then(|s| s.into_string().ok())
+      .unwrap_or_else(|| current_time.clone());
+
+    // tmp dir for extraction
+    let tmp_dir = tempfile::Builder::new()
+      .prefix(&format!("{}_{}_download", bin_name, current_time))
+      .tempdir()?;
+
+    // tmp directories are used to create backup of current application
+    // if something goes wrong, we can restore to previous state
+    let tmp_archive_path = tmp_dir.path().join(detect_archive_in_url(&url, &target));
+    let mut tmp_archive = File::create(&tmp_archive_path)?;
+
+    // set our headers
+    let mut headers = header::HeaderMap::new();
+    headers.insert(header::ACCEPT, "application/octet-stream".parse().unwrap());
+
+    // make sure we have a valid agent
+    if !headers.contains_key(header::USER_AGENT) {
+      headers.insert(
+        header::USER_AGENT,
+        "tauri/updater".parse().expect("invalid user-agent"),
+      );
+    }
+
+    // Set SSL for linux
+    set_ssl_vars!();
+
+    // Create our request
+    let resp = reqwest::Client::new()
+      .get(&url)
+      // wait 20sec for the firewall
+      .timeout(Duration::from_secs(20))
+      .headers(headers)
+      .send()
+      .await?;
+
+    // make sure it's success
+    if !resp.status().is_success() {
+      bail!(
+        crate::Error::Updater,
+        "Download request failed with status: {:?}",
+        resp.status()
+      )
+    }
+
+    tmp_archive.write_all(&resp.bytes().await?)?;
+
+    // Validate signature ONLY if pubkey is available in tauri.conf.json
+    if let Some(pub_key) = pub_key {
+      // we need an announced signature by the server
+      // if there is no signature, bail out.
+      if self.signature.is_none() {
```

Comment:
could be rewritten to `if let None = self.signature { ... }`  Actually surprised clippy didn't complain about this.